File: backend/app/schemas/job.py
# Purpose: Pydantic DTOs for Job endpoints including AI analysis fields.

from typing import Optional, Any
from pydantic import BaseModel, Field, model_validator
from datetime import datetime
from uuid import UUID


class JobCreate(BaseModel):
    title: str = Field(min_length=1, max_length=300)
    job_description: str = Field(min_length=1)
    free_text: Optional[str] = None
    icon: Optional[str] = Field(default=None, max_length=64)
    status: Optional[str] = Field(default="draft", max_length=32)
    additional_skills: Optional[list[str]] = None


class JobUpdate(BaseModel):
    title: Optional[str] = Field(default=None, max_length=300)
    job_description: Optional[str] = None
    free_text: Optional[str] = None
    icon: Optional[str] = Field(default=None, max_length=64)
    status: Optional[str] = Field(default=None, max_length=32)
    additional_skills: Optional[list[str]] = None
    analysis_json: Optional[dict[str, Any]] = None


class JobOut(BaseModel):
    id: UUID
    title: str
    job_description: str
    free_text: Optional[str]
    icon: Optional[str]
    status: str

    analysis_json: Optional[Any] = None
    analysis_model: Optional[str] = None
    analysis_version: Optional[int] = None
    ai_started_at: Optional[datetime] = None
    ai_finished_at: Optional[datetime] = None
    ai_error: Optional[str] = None

    created_at: datetime
    updated_at: datetime
    
    additional_skills: Optional[list[str]] = None

    @model_validator(mode='after')
    def extract_additional_skills(self):
        """Extract additional_skills from analysis_json if available."""
        if self.analysis_json and isinstance(self.analysis_json, dict):
            self.additional_skills = self.analysis_json.get('additional_skills')
        return self

    class Config:
        from_attributes = True


class JobListOut(BaseModel):
    items: list[JobOut]
    total: int

File: client/src/types/job.ts
// src/types/jobs.ts

export type LanguageLevel = 'basic' | 'conversational' | 'fluent' | 'native' | null;

export type CurrencyCode = 'ILS' | 'USD' | 'EUR' | null;

export type LanguageRequirement = {
  name: string;
  level: LanguageLevel;
};

export type SalaryRange = {
  min: number | null;
  max: number | null;
  currency: CurrencyCode;
};

export type TechStack = {
  languages: string[];
  frameworks: string[];
  databases: string[];
  cloud: string[];
  tools: string[];
  business: string[];
};

export type Experience = {
  years_min: number | null;
  years_max: number | null;
};

export type SecurityClearance = {
  mentioned: boolean;
  note: string | null;
};

export type JobAnalysis = {
  version: number;
  role_title: string | null;
  is_tech_role: boolean;
  organization: string | null;
  locations: string[];
  summary: string | null;

  responsibilities: string[];
  requirements: string[];

  skills: {
    must_have: string[];
    nice_to_have: string[];
  };
  
  additional_skills?: string[];

  experience: Experience;
  education: string[];

  salary_range: SalaryRange | null;

  security_clearance: SecurityClearance;

  tech_stack: TechStack;

  // Human languages with proficiency level
  languages: LanguageRequirement[];

  keywords: string[];
  evidence: string[];
};

/** ===== API models ===== */

export type ApiJob = {
  id: string;
  title: string;
  job_description: string;
  free_text: string | null;
  icon: string | null;
  status: string;
  created_at: string;
  updated_at: string;

  additional_skills: string[] | null;

  analysis_json: JobAnalysis | null;
  analysis_model: string | null;
  analysis_version: number | null;

  ai_started_at: string | null;
  ai_finished_at: string | null;
  ai_error: string | null;
};

export type CreateJobPayload = {
  title: string;
  job_description: string;
  free_text?: string;
  icon?: string;
  status?: string;
  additional_skills?: string[];
  analysis_json?: JobAnalysis;
};

export type UpdateJobPayload = {
  title?: string;
  job_description?: string;
  free_text?: string;
  icon?: string;
  status?: string;
  additional_skills?: string[];
  analysis_json?: JobAnalysis;
};

export type JobListResponse = {
  items: ApiJob[];
  total: number;
};

/** ===== Optional (legacy UI model) ===== */
export interface Job {
  id: string;
  title: string;
  description: string;
  freeText?: string;
  icon: string;
  postedAt: string;
  analysis: JobAnalysis | null;
  additionalSkills: string[];
}

export type JobDraft = Omit<Job, 'id' | 'postedAt' | 'analysis'> & { analysis_json?: JobAnalysis };
